### DISCUSSION TITLE
Expose rich structs and methods that use them?

### DISCUSSION THREAD

**Discussion by AndrewWinterman** [rabbitmq/amqp091-go#271] [Q&A]
Is there a reason structs like basicPublish are not exposed publicly? The existing interfaces are a bit hard to understand without an IDE or extensive comments.  In fact, I frequently find myself recreating them:

For clarity:

```
type basicPublish struct {
	Exchange   string
	RoutingKey string
	Mandatory  bool
	Immediate  bool
	Properties properties
	Body       []byte
}
```
and: 

```
	err := ch.BasicPublish(&BasicPublish{
		Exchange:   exchange,
		RoutingKey: key,
		Mandatory:  mandatory,
		Immediate:  immediate,
		Body:       msg.Body,
		Properties: properties{
			Headers:         msg.Headers,
			ContentType:     msg.ContentType,
			ContentEncoding: msg.ContentEncoding,
			DeliveryMode:    msg.DeliveryMode,
			Priority:        msg.Priority,
			CorrelationId:   msg.CorrelationId,
			ReplyTo:         msg.ReplyTo,
			Expiration:      msg.Expiration,
			MessageId:       msg.MessageId,
			Timestamp:       msg.Timestamp,
			Type:            msg.Type,
			UserId:          msg.UserId,
			AppId:           msg.AppId,
		},
	}); 
```

vs, for example

```
confirmation, err := ch.PublishWithDeferredConfirmWithContext(ctx, "application", "prospect", true, false,
		amqp091.Publishing{Body: []byte("hiiii")})

```

**Comment by lukebakken** [comment]
Hello, thanks for using this library.

Please note that Team RabbitMQ always uses GitHub Discussions for Q&A, not issues, which are for actionable work.

This library is a continuation of an old Go library (https://github.com/streadway/amqp) that made API decisions years ago.

In general, RabbitMQ client library APIs closely follow the AMQP 0.9.1 specification, for better or worse. We also try to minimize the amount of "sugar" in these libraries, preferring to keep them low-level and let the community provide more friendly alternatives built upon them.

@Zerpet @Gsantomaggio - anything to add here?

**Reply by Gsantomaggio** [reply]
> anything to add here?

Nope.

I'd add that creating a micro-personal library to wrap all these values is always possible.
